core/src/main/scala/neotypes/model.scala

```diff
@@ -0,0 +1,198 @@
+    def segments: List[Path.Segment]
+
+    def contains(node: Node): Boolean
+    def contains(relationship: Relationship): Boolean
+  }
+  object Path {
+    final case class EmptyPath(node: Node) extends Path {
+      override final val start: Node = node
+      override final val end: Node = node
+      override final val nodes: List[Node] = node :: Nil
+      override final val relationships: List[Relationship] = Nil
+      override final val segments: List[Segment] = Nil
+
+      override def contains(node: Node): Boolean =
+        this.node == node
+
+      override def contains(relationship: Relationship): Boolean =
+        false
+    }
+
+    final case class NonEmptyPath(segments: ::[Segment]) extends Path {
+      override final val start: Node =
+        segments.head.start
+
+      override def end: Node =
+        segments.last.end
+
+      override def nodes: List[Node] =
+        start :: segments.map(s => s.end)
+
+      override def relationships: List[Relationship] =
+        segments.map(s => s.relationship)
+
+      override def contains(node: Node): Boolean =
+        start == node || segments.exists(s => s.end == node)
+
+      override def contains(relationship: Relationship): Boolean =
+        segments.exists(s => s.relationship == relationship)
+    }
+
+    final case class Segment(start: Node, relationship: Relationship, end: Node)
+  }
+
+  /** Parent type of all Neo4j property types. */
+  sealed trait Value extends NeoType
+  object Value {
+    final case class ListValue[V <: SimpleValue](values: Iterable[V]) extends Value
+    sealed trait SimpleValue extends Value
+    sealed trait NumberValue extends SimpleValue
+    final case class Integer(value: Int) extends NumberValue
+    final case class Decimal(value: Double) extends NumberValue
+    final case class Str(value: String) extends SimpleValue
+    final case class Bool(value: Boolean) extends SimpleValue
+    final case class Bytes(value: ArraySeq[Byte]) extends SimpleValue
+    final case class Point(value: NeoPoint) extends SimpleValue
+    final case class Duration(value: NeoDuration) extends SimpleValue
+    sealed trait TemporalInstantValue extends SimpleValue
+    final case class LocalDate(value: JDate) extends TemporalInstantValue
+    final case class LocalTime(value: JTime) extends TemporalInstantValue
+    final case class LocalDateTime(value: JDateTime) extends TemporalInstantValue
+    final case class Time(value: JZTime) extends TemporalInstantValue
+    final case class DateTme(value: JZDateTime) extends TemporalInstantValue
+    final case object NullValue extends SimpleValue
+  }
+}
+
+/** Exceptions provided by this library. */
+object exceptions {
+  sealed abstract class NeotypesException(message: String, cause: Option[Throwable] = None) extends Exception(message, cause.orNull)
+
+  final case object TransactionWasNotCreatedException extends NeotypesException(message = "Couldn't create a transaction")
```

Comment:
Seems like these singleton exceptions should extend `NoStackTrace`, since their stack traces will not point to the callsite anyway.

Personally I am a bit skeptical about mixing case-class based ADTs with exceptions. `case class`es have a particular notion of equality that doesn't really take into account stacktraces etc. Even if you throw the same exception with the same message from the same place at two different points in time, are they _really_ equal? 🤔

Reply:
I disagree here, I think it's useful to have case class exceptions for the ability to pattern match on them in recovery combinators.

Reply:
You can get unapply without using case classes :) or their specific equality semantics

Reply:
That's true, I'd +1 _that_.

Reply:
> Seems like these singleton exceptions should extend NoStackTrace

I thought I did, but it seems I added it to the `ResultMapperException` hierarchy rather than the `NeotypesException`.

> Personally I am a bit skeptical about mixing case-class based ADTs with exceptions. case classes have a particular notion of equality that doesn't really take into account stacktraces etc

Fair.

> You can get unapply without using case classes

Yup :p